src/uint256.cpp

```diff
@@ -18,39 +18,31 @@ std::string base_blob<BITS>::GetHex() const
 }
 
 template <unsigned int BITS>
-void base_blob<BITS>::SetHex(const char* psz)
+void base_blob<BITS>::SetHex(std::string_view str)
```

Comment:
Not sure about `const`ing `string_view` parameters, could only find one pre-existing case of that pattern (`GetBuriedDeployment()`). Can sympathize with forcing introduction of new names over mutation. I like how you named the first digit `hi` (maybe `high` is more precise).

Ultimately, my modification of the function is meant to be a straightforward transformation of the original code. Since it has received a few ACKs already, I am going to put off changing it for now. But might draw some inspiration from your version if I need to change it for other reasons.

Reply:
We can get a reack, let's not focus on that.
I'm NACK 788fe9cc9ab979c5e14f544ae05dc46436892b81 -ing because of https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686336262 anyway.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1686328446

I don't really think it makes sense to make declare string views const, since the data they point to is already immutable, and the argument is passed by value so no changes to the view inside the function could affect the caller. I think the const just adds verbosity. Not 100% sure about this and haven't thought about it much, but would want to know a specific reason for using const here, or see some broader recommendation somewhere that would explain why to do this.

Reply:
> the data they point to is already immutable

The string part, but not the leading and trailing offsets.


The `const` signals the intention of the `string_view`.
By declaring it as `const std::string_view str` I'm announcing that its value will be constant throughout the method, right?
But if it's just `std::string_view str`, I can call `str.remove_prefix` or `str.remove_prefix`, which modifies the internals of the view, so after the call its internal state is different.

Reply:
@paplorinc I see `hi`/`lo` is consistent with `secp256k1_uint128`, so I retract my `high` thoughts.

It strikes me that the `hi` part of your suggestion is not the one being shifted `<< 4`. Maybe it should be called `lo`?

Reply:
you're absolutely right, edited the example!

Reply:
Let me try a different angle: if a method is inline, I would expect to be able to copy-paste its content to the call site directly (at least conceptually, to understand its effect).
If the method is pure in a functional sense (including that it doesn't even mutate its parameters), it's a lot easier to reason about its overall behavior.
Does this resonate more?

Reply:
re: https://github.com/bitcoin/bitcoin/pull/30436#discussion_r1688616750

> Let me try a different angle: if a method is inline, I would expect to be able to copy-paste its content to the call site directly (at least conceptually, to understand its effect). If the method is pure in a functional sense (including that it doesn't even mutate its parameters), it's a lot easier to reason about its overall behavior. Does this resonate more?

I actually learned something new about this looking at bc3139e5a5f1bf81e4bb53417157f0f2cef86716 from #30377: apparently our CI tools do not complain if you declare parameters `const` in the function definition but omit the `const` in the function declarations. I previously thought this would cause warnings or lint errors if definitions and declarations weren't exactly the same. So it seems like this provides a good compromise. You can avoid confusing callers by not including a `const` in the declaration that doesn't mean or do anything, while keeping const in the definition if you want to make it obvious that a parameter is not mutated.

I still think declaring const for function parameters passed by value doesn't make sense as a general rule, but it can make sense in certain cases for the reasons you explained.

I also learned there is a clang-tidy rule [`readability-avoid-const-params-in-decls`](https://clang.llvm.org/extra/clang-tidy/checks/readability/avoid-const-params-in-decls.html) that might be nice for us to enable.

Reply:
> function parameters passed by value doesn't make sense as a general rule

[Google cpp guide](https://google.github.io/styleguide/cppguide.html#Use_of_const ) mostly aligns with your view on this. However, I approach this from a more functional programming perspective, where I want to infer the possible state changes a method can perform based solely on its signature - or at least what it cannot do.

Thanks for following up on this, @ryanofsky!

Reply:
> I approach this from a more functional programming perspective, where I want to infer the possible state changes a method can perform based solely on its signature - or at least what it cannot do.

Yes, that's definitely good and desirable, but my point is that when you add `const` to a c++ function parameter that is passed by value (not passed by reference) the `const` **does not become part of the function signature**. It is completely ignored by callers and has no effect on what inputs can be passed to the function or what changes the function can make to those inputs (because c++ functions already cannot modify inputs which are passed by value).

The only effect of adding top-level const on function parameters passed by value is to make local argument variables inside the function definition constant instead of mutable. This can have benefits as well as drawbacks, just like functional programming sometimes has benefits over imperative programming, and sometimes has drawbacks. An important drawback that can happen when variables are immutable is it can lead to proliferation of similarly named variables containing slightly different values because it is not possible to replace existing values, leading to confusion between variables and bugs like https://github.com/bitcoin/bitcoin/pull/30386#discussion_r1683742656. I just think it is good idea to look at things on a case by case basis and not assume functional approaches are always or usually better.

Reply:
We're in 100% agreement here, thanks for detailing your view.

In this particular case my thinking was:
> initial code, parameter (copy of the original) is mutated locally, we have to keep track of state changes in our head to understand the method's behavior:
```C++
std::string_view RemovePrefixView2(std::string_view str, std::string_view prefix)
{
    if (str.starts_with(prefix)) {
        str.remove_prefix(prefix.size());
    }
    return str;
}
```
> adding const makes parameter mutations forbidden - we have to rethink the algo (i.e. code smell):
```C++
std::string_view RemovePrefixView3(const std::string_view str, const std::string_view prefix)
{
    if (str.starts_with(prefix)) {
        str.remove_prefix(prefix.size()); //  error: 'this' argument to member function 'remove_prefix' has type 'const std::string_view' (aka 'const basic_string_view<char>'), but function is not marked const
    }
    return str;
}
```
> we can return a copy of the view, knowing that `str` didn't change state throughout the method
```C++
std::string_view RemovePrefixView4(const std::string_view str, const std::string_view prefix)
{
    if (str.starts_with(prefix)) {
        return str.substr(prefix.size());
    }
    return str;
}
```
which enables:
```C++
std::string_view RemovePrefixView5(const std::string_view str, const std::string_view prefix)
{
    return str.starts_with(prefix)
         ? str.substr(prefix.size())
         : str;
}
```
At which point we might as well remove the `const` from the params, since the new impl is now trivial:
```C++
std::string_view RemovePrefixView5(std::string_view str, std::string_view prefix)
{
    return str.starts_with(prefix) ? str.substr(prefix.size()) : str;
}
```

Reply:
Makes sense that const can expose things even in simple cases and lead to insights and improvements like the one you are suggesting. In this case, I think "if starts with prefix call remove_prefix" might be more straightforward than "if starts with prefix call substr and return portion of the string after the prefix" but the difference is not great and I take your point in general const can have benefits in places like this.